compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp

```diff
@@ -62,93 +107,242 @@ FailureOr<bool> canFoldBasedOnHalfDmaCpy(
+  for (AMDAIE::NpuDmaWaitOp waitOp : waitOps) rewriter.eraseOp(waitOp);
+  return success();
+}
+
+/// Utility function to determine if a DMA wait operation can be folded into a
+/// a batch based on its half DMA copy operation.
+FailureOr<bool> canFoldByBatch(
+    Operation *batchParentOp, AMDAIE::NpuHalfDmaCpyNdOp npuHalfDmaCpyNdOp,
+    DenseSet<AMDAIE::ConnectionOp> &connectionOps,
+    DenseMap<DmaBdIdKey, DenseSet<uint32_t>> &dmaBdIdsMap) {
+  // Check if the current operation is in the same scope as the rest of the
+  // batch.
+  bool isSameScope = npuHalfDmaCpyNdOp->getParentOp() == batchParentOp;
+
+  // Retrieve the connection op.
+  std::optional<AMDAIE::ConnectionOp> maybeConnectionOp =
+      npuHalfDmaCpyNdOp.getConnectionOp();
+  if (!maybeConnectionOp) {
+    return npuHalfDmaCpyNdOp.emitOpError()
+           << "expected to operate on an `amdaie.connection`";
+  }
+  AMDAIE::ConnectionOp connectionOp = maybeConnectionOp.value();
+
+  // Retrieve the flow op.
+  std::optional<AMDAIE::FlowOp> maybeFlowOp = connectionOp.getFlowOp();
+  if (!maybeFlowOp) {
+    return connectionOp.emitOpError()
+           << "expected to operate on an `amdaie.flow`";
+  }
+  AMDAIE::FlowOp flowOp = maybeFlowOp.value();
+  bool isPacketFlow = flowOp.getIsPacketFlow();
+
+  // Retrieve the BD ID op.
+  std::optional<AMDAIE::BdIdOp> maybeBdIdOp = npuHalfDmaCpyNdOp.getBdIdOp();
+  if (!maybeBdIdOp) {
+    return npuHalfDmaCpyNdOp.emitOpError()
+           << "must have a BD ID op to lower to "
+              "`amdaie.npu.write_bd`";
+  }
+  AMDAIE::BdIdOp bdIdOp = maybeBdIdOp.value();
+
+  // Retrieve the tile op.
+  AMDAIE::TileOp tileOp =
+      dyn_cast_if_present<AMDAIE::TileOp>(bdIdOp.getTile().getDefiningOp());
+  if (!tileOp) {
+    return bdIdOp.emitOpError() << "must operate on an `amdaie.tile`";
+  }
+
+  bool isDuplicateConnection = connectionOps.contains(connectionOp);
+  uint32_t bdId = getConstantIndexOrAssert(bdIdOp.getValue());
+  bool isDuplicateBdId = llvm::any_of(dmaBdIdsMap, [&](const auto &entry) {
+    return entry.first.first == tileOp && entry.second.contains(bdId);
+  });
+
+  bool canFold = true;
+  // Can't fold wait op if:
+  // (1) the current connection op already occurs in the batch, or
+  // (2) the current BD ID on the same tile already occurs in the batch, or
+  // (3) the current operation is a packet flow, or
+  // (4) the batch is empty, or
+  // (5) the current operation is not in the same scope as the batch.
+  if (isDuplicateConnection || isDuplicateBdId || isPacketFlow ||
+      connectionOps.empty() || !isSameScope) {
+    // Clear the BD IDs for all the connections in the batch.
+    for (auto &entry : dmaBdIdsMap) {
+      ConnectionOp connectionOp = entry.first.second;
+      DenseSet<uint32_t> &bdIds = entry.second;
+      if (connectionOps.contains(connectionOp)) bdIds.clear();
+    }
+    connectionOps.clear();
+    canFold = false;
+  }
```

Comment:
I think it would be more clear to just return a bool from this function and do the clearing outside this function. This makes the function more standalone (single responsibility) and limited to what the name suggests (`canFoldByBatch`).

Reply:
And then you can also return early, for example:
```
bool isSameScope = npuHalfDmaCpyNdOp->getParentOp() == batchParentOp;
if (!isSameScope) return false;
```

Reply:
It is now separated into two functions, one just returns a bool

https://github.com/nod-ai/iree-amd-aie/blob/cc1ae9e027fc2a580421325de63725b90896cdfd/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp#L236

and the other does the update/clear

https://github.com/nod-ai/iree-amd-aie/blob/cc1ae9e027fc2a580421325de63725b90896cdfd/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFoldDmaWaits.cpp#L267